core/src/main/java/io/grpc/ProxyParameters.java

```diff
@@ -51,11 +72,48 @@ public boolean equals(Object o) {
     ProxyParameters that = (ProxyParameters) o;
     return Objects.equal(proxyAddress, that.proxyAddress)
         && Objects.equal(username, that.username)
-        && Objects.equal(password, that.password);
+        && Arrays.equals(password, that.password);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hashCode(proxyAddress, username, password);
+    return Objects.hashCode(proxyAddress, username, Arrays.hashCode(password));
+  }
+
+  /**
+   * Create a new builder.
+   */
+  public static Builder forAddress(InetSocketAddress proxyAddress) {
+    return new Builder(proxyAddress);
+  }
+
+  /**
+   * The helper class to build an Attributes instance.
+   */
+  public static final class Builder {
+
+    private final InetSocketAddress proxyAddress;
+    @Nullable
+    private String username;
+    @Nullable
+    private char[] password;
+
+    private Builder(InetSocketAddress proxyAddress) {
+      this.proxyAddress = Preconditions.checkNotNull(proxyAddress, "proxyAddress");
+    }
+
+    public Builder username(String username) {
+      this.username = username;
+      return this;
+    }
+
+    public Builder password(char[] password) {
```

Comment:
Everything on the propagation path, down to netty, would need to switch to `char[]` to achieve the goal of erasing the password after use. All components along the path would also need to coordinate to make sure the `byte[]` will not be used after it's erased, and I am a little worried that it may become too invasive of a change, while it's unclear to me how much more security it will bring. @ejona86 and @carl-mastrangelo what's your opinions?

Reply:
I agree with @zhangkun83 here.  It looks like the password is stored inside the Netty handler down the line, so deleting our ref is not sufficient.  Also, I think the bugs introduced by having a mutable array are more likely than the password field being leaked (this isn't C++, after all).

Reply:
Yeah, I agree too. Too much changes for an unclear benefit. I will undo the `char[]` changes on Monday